scipy/signal/_peak_finding.py

```diff
@@ -354,60 +358,9 @@ def peak_prominences(x, peaks, wlen=None):
     if wlen is not None and wlen < 3:
         raise ValueError('`wlen` must be at least 3')
 
-    # Prepare return arguments
-    prominences = np.zeros(peaks.size)
-    left_bases = np.zeros(peaks.size, dtype=int)
-    right_bases = np.zeros(peaks.size, dtype=int)
-
-    for i, peak in enumerate(peaks):
-        # If wlen is twice the size of x the symmetric window always covers
-        # the full x
-        if wlen is not None and wlen < x.size * 2:
-            wlen = int(wlen)
-            # Calculate window borders around the evaluated peak
-            wleft = peak - wlen // 2
-            wright = peak + wlen // 2
-            # Handle border cases
-            wleft = 0 if wleft < 0 else wleft
-            wright = x.size if x.size < wright else wright
-            # Use slice for prominence calculation
-            window = x[wleft:wright]
-            # Correct peak position in x
-            peak -= wleft
-        else:
-            # Use full x for prominence calculation
-            window = x
-            wleft = 0
-
-        # Positions where window is larger than current peak height
-        greater_peak = np.where(window > window[peak])[0]
-
-        try:
-            # Nearest position to the left of peak with
-            # window[left] > window[peak]
-            left = greater_peak[greater_peak < peak].max()
-        except ValueError:
-            left = 0
-        try:
-            # Nearest position to right of peak with
-            # window[right] > window[peak]
-            right = greater_peak[greater_peak > peak].min()
-        except ValueError:
-            right = None
-
-        # Base indices to the left and right of peak in window
-        left_bases[i] = window[left:peak].argmin() + left
-        right_bases[i] = window[peak:right].argmin() + peak
-
-        # Calculate lowest contour and its vertical distance to peak
-        lowest_contour = max(window[left_bases[i]], window[right_bases[i]])
-        prominences[i] = window[peak] - lowest_contour
-
-        # Correct window offset
-        left_bases[i] += wleft
-        right_bases[i] += wleft
-
-    return prominences, left_bases, right_bases
+    peaks = np.asarray(peaks, dtype=np.intp)
```

Comment:
You already call `np.asarray(peaks)` on line 343.

Reply:
Yeah, but in that case i call it without the `dtype=np.intp` argument. 
In the first case I do it in order to test for peaks size and dimension. If I would add the typecast here the check `np.issubdtype(peaks.dtype, np.integer)` after this would be meaningless.

I think this is a little ugly but I couldn't find a more elegant solution that checks for integerness of peaks and ensures that the correct int-size is used on all architectures.

Reply:
How about
```
peaks = np.asarray(peaks).astype(np.intp, order='C', casting='safe', subok=False, copy=False)
```

on line 343. This would raise an error if `peaks.dtype` isn't safely castable to `intp` (essentially the same as your manual check) and also makes sure it's stored contiguously.